### PR TITLE
Parallel ESMF builds

### PR DESCRIPTION
**Pull request by bekozi**
Adding parallel builds for linux and osx. Not convinced osx is going to work.

### CODE REVIEW

**Comment by ocefpaf** [comment]
@bekozi you'll some of te changes I added in #8. Let me merged that one first and you can rebase this.

**Comment by ocefpaf** [comment]
Please rebase against master so we can get the `perl` changes in. (Sorry for being slow to add that.)

>  Not convinced osx is going to work.

If it does not let's get only the Linux changes and then open another PR with the failing OS X try so we can ask Mac owners for help :wink:

**Comment by bekozi** [comment]
Changes are merged.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -7,17 +7,21 @@ source:
...
+    number: 3
     skip: True  # [win]
 
 requirements:
     build:
         - gcc
         - netcdf-fortran 4.4*
         - perl
+        - openmpi # [osx]
+        - mpich2 # [linux]
```
Comment (ocefpaf) [inline recipe/meta.yaml]: The linter request two spaces before the `#` :stuck_out_tongue_winking_eye:

**Comment by bekozi** [comment]
I'm not sure what the issue is here. We build parallel for Linux successfully with our CentOS docker image: https://github.com/NESII/conda-recipes-nesii/blob/master/Dockerfile-CentOS6-Builder. We do require system `gcc` and `gfortran`, however. The default channel `mpich2` is used.

I don't have a good sense of what's different about the `conda-forge` build environment. I can test with the `condaforge/linux-anvil` docker image at some point, but I fear the result will be changes to the default channel `mpich2` or requiring a `conda-forge` build of `mpich2` or `openmpi`. Any idea why the parallel Linux build worked under the old system (`IOOS`) and not here?

**Comment by ocefpaf** [comment]
In the `IOOS` builds (probably like your own) we used `gfortran` from the CentOS image. In `conda-forge` we avoid that for now b/c it would require all users to install the same `gfortran` (not that using the default channel `gcc` is a much better option though...).

I will investigate if this is just a case of adding a symlink to `f95` or something more complicated.

**Comment by bekozi** [comment]
> I will investigate if this is just a case of adding a symlink to f95 or something more complicated.

A little nostalgia for `mpich2`: https://github.com/ioos/conda-recipes/issues/676#issuecomment-166982432

**Comment by ocefpaf** [comment]
> A little nostalgia for `mpich2`: https://github.com/ioos/conda-recipes/issues/676#issuecomment-166982432


Full circle :smile: 

I am traveling later today (back home after almost a month away) and I will take a look at what we can do. Ideally we should have builds for `mpich2` and `openmpi` in conda-forge.

**Comment by bekozi** [comment]
Safe travels and welcome home!

**Comment by bekozi** [comment]
@ocefpaf Is it possible to skip Travis-CI builds with magic keywords in the commit message?

**Comment by ocefpaf** [comment]
> @ocefpaf Is it possible to skip Travis-CI builds with magic keywords in the commit message?

In the past `[skip ci]` did not work on `CircleCI`, so that would do. But if you want to skip only `parallel` builds need to use `[linux]` in the dependencies for the parallel builds and use an `if-clause` in the `build.sh`.

**Comment by bekozi** [comment]
> In the past [skip ci] did not work on CircleCI, so that would do.

Worked. For this testing I'm not interested in the `osx` builds as no MPI options are available...

**Comment by bekozi** [comment]
@ocefpaf Would it be difficult to support the 1.4.x version of `mpich`? `esmf` supports `mpich3` but only as a dev feature. It is complex enough that putting a patch together would be difficult. Considering the backward-incompatible changes in the new `mpich` version, perhaps supporting an earlier version is okay? Let me know if I can help out.

**Comment by rokuingh** [comment]
According to our new findings with mpi4py not working properly with mpich
would it be possible to use openmpi instead?

**Comment by ocefpaf** [comment]
> Would it be difficult to support the 1.4.x version of `mpich`? 

I can try a recipe fir `1.4.x` soon.

> would it be possible to use `openmpi` instead?

The default channel has `openmpi` on `OS X` so we can use that as a starting point.  I am stick with my day job duties for a while, but I will try to get back to this soon. (If not ping me again :grimacing:)

**Comment by bekozi** [comment]
> According to our new findings with mpi4py not working properly with mpich would it be possible to use openmpi instead?

Thanks for the reminder @rokuingh. @ocefpaf, would `conda-forge` be interested in supporting `openmpi`? `mpich` has issues with `mpi.spawn`.

>  I am stick with my day job duties for a while, but I will try to get back to this soon.

You have a day job too?? Geez. :smile:

**Comment by ocefpaf** [comment]
> would `conda-forge` be interested in supporting openmpi?

Yes. And I have a recipe that is almost working. Just having trouble finding those `.txt` at test time when calling `mpicc` for example.

> You have a day job too?? Geez.

Yeah... Day jobs always interfering with fun work :smile:

**Comment by bekozi** [comment]
> Yes. And I have a recipe that is almost working. Just having trouble finding those .txt at test time when calling mpicc for example.

Fun. Do you have a link to your current recipe?

**Comment by ocefpaf** [comment]
> Do you have a link to your current recipe?

It is here in case you want to finish it: https://github.com/ocefpaf/staged-recipes/tree/openmpi/recipes/openmpi

**Comment by bekozi** [comment]
Closing this to start anew.

**Pull request closed without merging**